api/v1/server/handlers/v1/webhooks/receive.go

```diff
@@ -0,0 +1,417 @@
+			}
+		}
+
+		if !signaturesMatch(signature, expectedSignature) {
+			return false, &ValidationError{
+				Code:      Http403,
+				ErrorText: "invalid HMAC signature",
+			}
+		}
+	case sqlcv1.V1IncomingWebhookSourceNameGITHUB:
+		fallthrough
+	case sqlcv1.V1IncomingWebhookSourceNameGENERIC:
+		switch webhook.AuthMethod {
+		case sqlcv1.V1IncomingWebhookAuthTypeBASIC:
+			username, password, ok := request.BasicAuth()
+
+			if !ok {
+				return false, &ValidationError{
+					Code:      Http403,
+					ErrorText: "missing or invalid authorization header",
+				}
+			}
+
+			decryptedPassword, err := w.config.Encryption.Decrypt(webhook.AuthBasicPassword, "v1_webhook_basic_auth_password")
+
+			if err != nil {
+				return false, &ValidationError{
+					Code:      Http500,
+					ErrorText: fmt.Sprintf("failed to decrypt basic auth password: %s", err),
+				}
+			}
+
+			if username != webhook.AuthBasicUsername.String || password != string(decryptedPassword) {
+				return false, &ValidationError{
+					Code:      Http403,
+					ErrorText: "invalid basic auth credentials",
+				}
+			}
+		case sqlcv1.V1IncomingWebhookAuthTypeAPIKEY:
+			apiKey := request.Header.Get(webhook.AuthApiKeyHeaderName.String)
+
+			if apiKey == "" {
+				return false, &ValidationError{
+					Code:      Http403,
+					ErrorText: fmt.Sprintf("missing or invalid api key header: %s", webhook.AuthApiKeyHeaderName.String),
+				}
+			}
+
+			decryptedApiKey, err := w.config.Encryption.Decrypt(webhook.AuthApiKeyKey, "v1_webhook_api_key")
+
+			if err != nil {
+				return false, &ValidationError{
+					Code:      Http500,
+					ErrorText: fmt.Sprintf("failed to decrypt api key: %s", err),
+				}
+			}
+
+			if apiKey != string(decryptedApiKey) {
+				return false, &ValidationError{
+					Code:      Http403,
+					ErrorText: fmt.Sprintf("invalid api key: %s", webhook.AuthApiKeyHeaderName.String),
+				}
+			}
+		case sqlcv1.V1IncomingWebhookAuthTypeHMAC:
+			signature := request.Header.Get(webhook.AuthHmacSignatureHeaderName.String)
+
+			if signature == "" {
+				return false, &ValidationError{
+					Code:      Http403,
+					ErrorText: fmt.Sprintf("missing or invalid signature header: %s", webhook.AuthHmacSignatureHeaderName.String),
+				}
+			}
+
+			decryptedSigningSecret, err := w.config.Encryption.Decrypt(webhook.AuthHmacWebhookSigningSecret, "v1_webhook_hmac_signing_secret")
+
+			if err != nil {
+				return false, &ValidationError{
+					Code:      Http500,
+					ErrorText: fmt.Sprintf("failed to decrypt HMAC signing secret: %s", err),
+				}
+			}
+
+			algorithm := webhook.AuthHmacAlgorithm.V1IncomingWebhookHmacAlgorithm
+			encoding := webhook.AuthHmacEncoding.V1IncomingWebhookHmacEncoding
+
+			expectedSignature, err := computeHMACSignature(webhookPayload, decryptedSigningSecret, algorithm, encoding)
+
+			if err != nil {
+				return false, &ValidationError{
+					Code:      Http500,
+					ErrorText: fmt.Sprintf("failed to compute HMAC signature: %s", err),
```

Comment:
Do we want to return unauthorized here instead? Just thinking that we might not want to get swarmed with 500s

Reply:
good catch 👍